keps/sig-api-machinery/4420-retry-generate-name/README.md

```diff
@@ -0,0 +1,541 @@
+
+No
+
+### Troubleshooting
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+The only know failure modes are pathological. E.g. A runaway system creating
+unbounded resources using generateName could eventually trigger amplified load
+on webhooks with this feature.
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+Look at the number of resources (using kube-state-metrics or similar). If there
+exists a very large number of resources (millions) of a particular resource kind
+and they all share the same generateName prefix, this might be triggering
+high retry load.
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+This code is in a critical system codepath, if anything is wrong with it, it could cause
+severe problems, up to and including data corruption and loss.
+
+## Alternatives
+
+There are many alternatives that we considered including:
+
+- Increasing the length of the random suffix appended to generateName when generating
+  a name. This would be a breaking change in two ways:
+  - There is a 63 char name length limit. to guarantee a 5 char random suffix
+    length, generateName is trimmed if its length is >58 chars. If the random
+    suffix length were to be increased, generateName would need to trimmed to a
+    shorter length, breaking users that depend on 58 chars of generateName being
+    preserved. (Increasing the max lengh of name would also be a breaking
+    change).
+  - There are major downstream projects that use "{5}" regexes to match the
+    random suffix.
+  - Note that the selected design is statistically as good as increasing the
+    characters appended from 5 to 11.
+- After generating a name, check if the name exists in etcd and, if it does,
+  generate another name.  See https://github.com/kubernetes/kubernetes/compare/master...jpbetz:kubernetes:retry-generate-name
+  The downside to this approach is that it adds a round-trip to etcd for
+  all create requests that use generateName.
+- Add a "retry filter" to the apiserver's filter chain to retry the entire
+  request. This would reinvoke the entire admission chain and increase
+  the latency of request handling more significantly than the other approaches
+  considered.
+- Introduce automatic retry logic into the client.
```

Comment:
We could also do both; the client could signify its willingness to automatically retry (eg, with a special query string on the **create** request).

Reply:
I like it. I'll make a note.